recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+            )
+            if self.hparams.use_wandb:
+                wandb.log(
+                    {
+                        "valid_acuracy": stage_loss,
+                    },
+                    step=self.optimizer_step,
+                )
+
+    def update_average(self, loss, avg_loss):
+        if avg_loss == 0.0:
+            avg_loss = loss.item()
+        else:
+            avg_loss = 0.99 * avg_loss + 0.01 * loss.item()
+        return avg_loss
+
+
+def sample_negatives(y, num_neg):
+    """
+       y is output of feature extractor
+    """
+    B, T, C = y.shape
+    high = T - 1
+    with torch.no_grad():
+        targets = torch.arange(T).unsqueeze(-1).expand(-1, num_neg).flatten()
+        neg_indcs = torch.randint(low=0, high=high, size=(B, T * num_neg))
+        # negative should not be target and to make distribution uniform shift all >
+        neg_indcs[neg_indcs >= targets] += 1
+
+    neg_indcs = neg_indcs + torch.arange(B).unsqueeze(1) * high
+    y = y.view(-1, C)
+    negs = y[neg_indcs.view(-1)]
+    negs = negs.view(B, T, num_neg, C).permute(2, 0, 1, 3)  # to N, B, T, C
+    return negs
+
+
+def dataio_prepare(hparams):
+    data_folder = hparams["data_folder"]
+
+    train_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["train_csv"], replacements={"data_root": data_folder},
+    )
+
+    # we sort training data to speed up training and get better results.
+    train_data = train_data.filtered_sorted(
+        sort_key="duration",
+        key_max_value={"duration": hparams["avoid_if_longer_than"]},
+        key_min_value={"duration": hparams["avoid_if_shorter_than"]},
+    )
+
+    valid_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["valid_csv"], replacements={"data_root": data_folder},
+    )
+    # We also sort the validation data so it is faster to validate
+    valid_data = valid_data.filtered_sorted(sort_key="duration")
+
+    datasets = [train_data, valid_data]
+
+    @sb.utils.data_pipeline.takes("wav")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(wav):
+        info = torchaudio.info(wav)
+        sig = sb.dataio.dataio.read_audio(wav)
+        assert sig.dim() == 1, sig.dim()
+        resampled = torchaudio.transforms.Resample(
+            info.sample_rate, hparams["sample_rate"],
+        )(sig)
+        with torch.no_grad():
+            resampled = F.layer_norm(resampled, resampled.shape)
+        return resampled
+
+    sb.dataio.dataset.add_dynamic_item(datasets, audio_pipeline)
+
+    sb.dataio.dataset.set_output_keys(datasets, ["id", "sig"])
+
+    train_sampler = DynamicBatchSampler(
+        train_data,
+        hparams["seconds_per_batch"],
+        num_buckets=70,
+        length_func=lambda x: x["duration"],
+        shuffle=True,
+        batch_ordering="random",
+    )
+
+    W2VPaddedBatch_partial = functools.partial(W2VPaddedBatch,
```

Comment:
What is that ?

Reply:
So I noticed a significant speedup from moving the mask creation into the dataloader. This is what huggingface does and is I suspect the major reason why you find it to be faster than fairseq.

But for the mask creation I need to know some information about the model like the size and number of kernels+strides. So I need to pass that information to a custom collate class (`W2VPaddedBatch`) of the dataloader. `functools.partial` allows me to pass that information to the class without actually initialising it (by setting some specific arguments)